android/src/main/kotlin/com/hypertrack/sdk/flutter/common/Serialization.kt

```diff
@@ -0,0 +1,172 @@
+}
+
+internal fun serializeLocation(location: Location): Map<String, Any> {
+    return mapOf(
+        KEY_TYPE to TYPE_LOCATION,
+        KEY_VALUE to mapOf(
+            KEY_LATITUDE to location.latitude,
+            KEY_LONGITUDE to location.longitude
+        )
+    )
+}
+
+internal fun serializeLocationError(locationError: LocationError): Map<String, Any> {
+    return when (locationError) {
+        NotRunning -> {
+            mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_NOT_RUNNING)
+        }
+        Starting -> {
+            mapOf(KEY_TYPE to TYPE_LOCATION_ERROR_STARTING)
+        }
+        is Errors -> {
+            mapOf(
+                KEY_TYPE to TYPE_LOCATION_ERROR_ERRORS,
+                KEY_VALUE to locationError.errors.map { serializeHypertrackError(it) }
+            )
+        }
+    }
+}
+
+internal fun serializeHypertrackError(error: HyperTrackError): Map<String, String> {
+    return mapOf(
+        KEY_TYPE to TYPE_HYPERTRACK_ERROR,
+        KEY_VALUE to error.name
+    )
+}
+
+internal fun deserializeAvailability(isAvailable: Map<String, Any>): Result<Boolean> {
+    return parse(isAvailable) {
+        it.assertValue<String>(key = KEY_TYPE, value = TYPE_AVAILABILITY)
+        it.get<Boolean>(KEY_VALUE).forceUnwrap()
+    }
+}
+
+internal fun deserializeGeotagData(map: Map<String, Any>): Result<Geotag> {
+    return parse(map) {
+        val data = it.get<Map<String, Any>>(KEY_GEOTAG_DATA).forceUnwrap()
+        Geotag(data)
+    }
+}
+
+private fun <T> parse(
+    source: Map<String, Any>,
+    block: (ParsingHandle) -> T
+): Result<T> {
+    val handle = ParsingHandle(source);
+    return try {
+        if (handle.exceptions.isNotEmpty()) {
+            Failure(ParsingExceptions(source, handle.exceptions))
+        } else {
+            Success(block.invoke(handle))
+        }
+    } catch (e: Exception) {
+        Failure(
+            if (handle.exceptions.isNotEmpty()) {
+                ParsingExceptions(source, handle.exceptions + listOf(e))
+            } else {
+                e
+            }
+        )
+    }
+}
+
+private class ParsingHandle(
+    private val source: Map<String, Any>,
+) {
+    private val _exceptions = mutableListOf<Exception>()
+    val exceptions: List<Exception> = _exceptions
+
+    inline fun <reified T> get(
+        key: String
+    ): Result<T> {
+        return try {
+            Success(source[key]!! as T)
+        } catch (e: Exception) {
+            Failure(ParsingException(key, e).also {
+                _exceptions.add(it)
+            })
+        }
+    }
+
+    inline fun <reified T> assertValue(
+        key: String,
+        value: Any
+    ) {
+        if (source[key] != value) {
+            _exceptions.add(Exception("Assertion failed: $key != $value"))
+        }
+    }
+}
+
+private data class ParsingExceptions(
+    val source: Any,
+    val exceptions: List<Exception>
+) : Throwable(exceptions.joinToString("\n").let {
+    "Invalid input:\n\n${source}\n\n${it}"
```

Comment:
I'm not convinced message should be multiline with multi newlines one after another, it is supposed to be short, not a format that looks very debugish. You may have issues with printing that in different exception handlers.

Reply:
resolved in call